Добавление кнопки "Откликнуться" под сообщение с заданием

eugemos opened the pull request:
### Что сделано
1. Реализована функциональность в соответствии с условиями задачи #545.
2. Отдельный репозиторий и сервис для модели TaskResponseVolunteer создавать не стал. Поместил необходимые методы в репозиторий для модели ExternalSiteUser и соответствующий сервис. У нас уже есть аналогичный случай с моделью UsersCategories, там все методы размещены в репозитории и сервисах для модели User. Я решил не отклоняться от этой линии.
3. Пришлось также исправить обработчики эндпойнтов `task.actualize_tasks` и `task.create_update_task`, так как они использовали функцию `get_task_info_keyboard`, которую я изменил. Эту функцию стало невозможно там использовать, так как она теперь требует в качестве параметра пользователя сайта, потому что клавиатура под сообщением теперь для каждого пользователя сайта должна быть своя, в зависимости от того, был ли у него отклик на данную задачу или нет. Теперь в метод `TelegramNotificationService.send_messages_to_subscribed_users` передаётся шаблон сообщения, который позже преобразуется в телеграм-сообщение. Соответственно, добавлены два новых класса: `TaskInfoMessageTemplate` и `TelegramMessageTemplate`, и метод `TelegramNotification.send_messages_by_template`. Незарегистрированным на сайте пользователям отправка сообщения заблокирована.

### Как тестировал
Локально. Использовал страницу документации API. Для просмотра БД использовал DBeaver.
#### Основная функциональность
Проверял, что под сообщением с информацией о задании появилась кнопка "Откликнуться"/"Отменить отклик" и что она ведёт себя, как указано в задаче #545.
1. Заполнил БД тестовыми данными без фейковых пользователей.
2. Через API создал пользователя сайта, связал с ним пользователя Телеграм (`/start id_hash`).
4. Вывел список актуальных заданий (`Меню/Посмотреть актуальные задания`). Затем стартовал бот ещё раз (`/start`) и вывел ещё один список актуальных заданий.
5. Проверил наличие кнопки "Откликнуться" под описаниями задач. Во втором списке проверил как работает эта кнопка - при её нажатии она превращается в кнопку "Отметить отклик", а в БД появляется соответствующая запись в таблице `task_response_volunteer `. Аналогично проверил работу кнопки  "Отметить отклик".
6. Первый список использовал для проверки реакции на устаревшие кнопки путём дублирования в нём действий, выполненных во втором списке.
7. Через API заархивировал одну из задач из списка и проверил поведение устаревших кнопок для этого случая.
8. С помощью другого устройства привязал к пользователю сайта другого пользователя Телеграм и проверил, что первый пользователь Телеграм при нажатии на устаревшие кнопки получает предложение зарегистрироваться.
#### Исправленные эндпойнты
Проверял, что пользователь получает надлежащее уведомление в том случае, если изменена или создана задача в одной из его категорий, при условии, что он зарегистрирован на сайте и подписан на уведомления в боте.
1. Использовал двух разных пользователей, у которых была одна общая категория, а остальные отличались, например, [11, 13] и [12, 13]. Пользователи подписаны на задания в боте.
2. Создал через API (`POST: /api/tasks`) набор новых задач во всех вышеперечисленных категориях + одна в категории, которой нет ни у одного из пользователей. Проверил правильность полученных пользователями сообщений.
3. Затем вносил изменения в этот набор задач как списком  (`POST: /api/tasks`), так и по отдельности  (`POST: /api/task`) и проверял правильность получаемых пользователями сообщений. 
5. Отменял у пользователей подписку в боте и проверял, что они перестали получать сообщения при изменении задач в их категориях.
6. Отменял регистрацию пользователя вручную через DBeaver и проверял, что он перестал получать сообщения при изменении задач в его категориях.
Примечание.
После ручного изменения БД бота следует перезапускать, иначе он может работать некорректно.

eugemos committed:
Добавлена кнопка откликнуться под кратким описанием задачи.

eugemos committed:
Реализован отклик пользователя на задачу и отмена отклика.

gorskyolga commented on src/bot/handlers/tasks.py:
```diff
@@ -48,7 +48,7 @@ async def view_task_callback(
             parse_mode=ParseMode.HTML,
             disable_web_page_preview=True,
-            reply_markup=get_task_info_keyboard(task),
+            reply_markup=await get_task_info_keyboard(task, ext_site_user),
```
get_task_info_keyboard также используется в `src\api\endpoints\tasks.py` в actualize_tasks и create_update_task. Нужно подтянуть изменения из ветки debelop, create_update_task была недавно добавлена.

gorskyolga commented on src/bot/handlers/tasks.py:
```diff
@@ -70,5 +70,41 @@ async def show_next_tasks(update: Update, context: CallbackContext, page_number:
+@registered_user_required
+async def respond_to_task_callback(
+    update: Update,
+    context: CallbackContext,
```
Заметила, что везде в хэндлерах `context: ContextTypes.DEFAULT_TYPE` и только в `src/bot/handlers/tasks.py` используется `context: CallbackContext`. Луше поменять на `context: ContextTypes.DEFAULT_TYPE` для единообразия.

gorskyolga commented on src/core/db/repository/external_site_user.py:
```diff
@@ -21,3 +21,50 @@ async def get_or_create_by_id_hash(self, id_hash: str) -> tuple[ExternalSiteUser
+
+    async def create_user_response_to_task(self, site_user: ExternalSiteUser, task: Task) -> bool:
+        """Создаёт отклик заданного пользователя на заданную задачу и возвращает True,
+        если такого отклика ещё нет в БД, иначе просто возвращает False.
```
Мне кажется, как-то сложно читается докстринг. Может его так изменить здесь и в сервисе:

Создаёт отклик заданного пользователя на заданную задачу и возвращает True.
Если такой отклик уже есть в БД, просто возвращает False.

gorskyolga commented on src/core/db/repository/external_site_user.py:
```diff
@@ -21,3 +21,50 @@ async def get_or_create_by_id_hash(self, id_hash: str) -> tuple[ExternalSiteUser
+    async def delete_user_response_to_task(self, site_user: ExternalSiteUser, task: Task) -> bool:
+        """Удаляет отклик заданного пользователя на заданную задачу и возвращает True,
+        если такой отклик есть в БД, иначе просто возвращает False.
+        """
```
Мне кажется, сложно читается докстринг. Может его так изменить здесь и в сервисе:

Удаляет отклик заданного пользователя на заданную задачу и возвращает True.
Если такого отклика уже нет в БД, просто возвращает False.

gorskyolga commented on src/core/db/repository/external_site_user.py:
```diff
@@ -21,3 +21,50 @@ async def get_or_create_by_id_hash(self, id_hash: str) -> tuple[ExternalSiteUser
+        если такого отклика ещё нет в БД, иначе просто возвращает False.
+        """
+        await self._session.commit()
+        async with self._session.begin():
```
Мне кажется, нужно использовать декоратор auto_commit из `src.core.utils.py` вместо этих двух строчек.

gorskyolga commented on src/core/db/repository/external_site_user.py:
```diff
@@ -21,3 +21,50 @@ async def get_or_create_by_id_hash(self, id_hash: str) -> tuple[ExternalSiteUser
+        если такой отклик есть в БД, иначе просто возвращает False.
+        """
+        await self._session.commit()
+        async with self._session.begin():
```
Мне кажется, нужно использовать декоратор auto_commit из `src.core.utils.py` вместо этих двух строчек.

eugemos committed:
Merge branch 'develop' into feature/add-response-button-for-task-545

eugemos committed:
Merge branch 'develop' into feature/add-response-button-for-task-545

eugemos committed:
Исправлена отправка телеграм-сообщений с информацией о задаче
из обработчиков эндпойнтов tasks.actualize_tasks() и task.create_update_task().

eugemos committed:
Заменил CallbackContext на ContextTypes.DEFAULT_TYPE.

eugemos committed:
Поправил докстринги.

eugemos committed:
Применил декоратор auto_commit.

eugemos committed:
Добавил докстринги.

Xalgina commented:
> У сообщений с заданиями есть кнопка "Посмотреть задание". Нужно добавить под неё кнопку "Откликнуться".

**Протестировано. ОК, кнопка есть**

> Если отклика волонтера на задание нет, то текст на кнопке "Откликнуться".

**Протестировано. ОК**

> Если отклик волонтера на задание есть, то текст на кнопке "Отменить отклик".

**Протестировано. ОК**

> Должна проверяться авторизация пользователя в боте. Пользователь Телеграм, у которого нет авторизации (связи с external_user, роли у external_user и т.д.), должен получать сообщение с требованием авторизоваться. 

**Протестировано. ОК**

> Если задание заархивировано, то должен отображаться поп-апп с текстом "Задание уже отдано в работу другому волонтеру".

**Протестировано. ОК**

> Если волонтер нажимает на кнопку "Откликнуться", то должна создаваться запись об отклике в таблице task_response_volunteer и текст на кнопке должен меняться на "Отменить отклик".

**Протестировано. ОК**

> Если волонтер нажимает на кнопку "Отменить отклик", то запись об отклике должна удаляться из таблицы task_response_volunteer и текст на кнопке должен меняться на "Откликнуться".

**Протестировано. ОК**


> при попытке откликнуться на задание, по которому волонтер уже откликнулся, нужно отображать поп-апп с текстом "Ты уже откликнулся на это задание";

**Протестировано. ОК**

> при попытке отменить отклик на задание, по которому отклик уже был отменен, нужно отображать поп-апп с текстом "Ты уже отменил отклик на это задание".

**Протестировано. ОК**

> При этом кнопки должны меняться на правильные.

**Протестировано. ОК**

gorskyolga committed:
Merge branch 'develop' into feature/add-response-button-for-task-545

gorskyolga commented on src/core/db/repository/external_site_user.py:
```diff
@@ -24,6 +25,51 @@ async def get_or_create_by_id_hash(self, id_hash: str) -> tuple[ExternalSiteUser
+                external_site_user_id=site_user.id,
+                task_id=task.id,
+            )
+            self._session.add(response)
```
Я задумалась, не лучше ли будет вместо self._session.add и auto_commit использовать метод create из AbstractRepository.

eugemos replied:
С одной стороны, конечно, лучше воспользоваться готовым методом, что бы все объекты создавались по одной схеме независимо от их типа. А с другой стороны есть вопрос: а законно ли использовать `ExternalSiteUserRepository` для создания объекта типа `TaskResponseVolunteer`? Я именно про законность, технически-то это будет работать. Дело тут в том, что  `AbstractRepository` - обобщённый тип и при применении, по идее, должен конкретизироваться типом модели, то есть должно было быть
```
class ExternalSiteUserRepository(AbstractRepository[ExternalSiteUser]):
  ...
```
а не 
```
class ExternalSiteUserRepository(AbstractRepository):
  ...
```
как у нас, и должен применяться линтер, который бы проверял соответствие типов, и  "не позволял" бы использовать методы `AbstractRepository[ExternalSiteUser]` для работы с `TaskResponseVolunteer`. Но у нас сейчас нет ни того ни другого.

И тут надо определиться, каковы наши намерения относительно линтера и аннотаций типов. Если мы хотели бы по-честному придерживаться заданных типов, тогда надо использовать линтер и на него ориентироваться в вопросах, что можно и что нельзя. Если же нет, тогда, конечно, мы можем позволить себе ~~не соблюдать диету~~ больше, но это плохой путь с моей точки зрения.

gorskyolga replied:
Добавила в бэклог.

gorskyolga commented on src/core/db/repository/external_site_user.py:
```diff
@@ -24,6 +25,51 @@ async def get_or_create_by_id_hash(self, id_hash: str) -> tuple[ExternalSiteUser
+        """
+        response = await self.get_user_response_to_task_or_none(site_user, task)
+        if response is not None:
+            await self._session.delete(response)
```
А здесь соответственно вместо self._session.delete и auto_commit использовать метод remove из AbstractRepository?

gorskyolga replied:
Добавила в бэклог.
